src/table.js

```diff
@@ -445,64 +450,168 @@ Table.prototype.createReadStream = function(options) {
-    tableName: this.id,
-    objectMode: true,
-  };
+  let rowKeys;
+  let ranges = options.ranges || [];
+  let filter;
+  let rowsLimit;
+  let rowsRead = 0;
+  let numRequestsMade = 0;
 
   if (options.start || options.end) {
-    options.ranges.push({
+    ranges.push({
       start: options.start,
       end: options.end,
     });
   }
 
-  if (options.prefix) {
-    options.ranges.push(Table.createPrefixRange_(options.prefix));
+  if (options.keys) {
+    rowKeys = options.keys;
   }
 
-  if (options.keys || options.ranges.length) {
-    reqOpts.rows = {};
-
-    if (options.keys) {
-      reqOpts.rows.rowKeys = options.keys.map(Mutation.convertToBytes);
-    }
-
-    if (options.ranges.length) {
-      reqOpts.rows.rowRanges = options.ranges.map(function(range) {
-        return Filter.createRange(range.start, range.end, 'Key');
-      });
-    }
+  if (options.prefix) {
+    ranges.push(Table.createPrefixRange_(options.prefix));
   }
 
   if (options.filter) {
-    reqOpts.filter = Filter.parse(options.filter);
+    filter = Filter.parse(options.filter);
   }
 
   if (options.limit) {
-    reqOpts.rowsLimit = options.limit;
+    rowsLimit = options.limit;
   }
-  const chunkTransformer = new ChunkTransformer({decode: options.decode});
-  const stream = pumpify.obj([
-    this.requestStream(grpcOpts, reqOpts),
-    chunkTransformer,
-    through.obj(function(rowData, enc, next) {
-      if (stream._ended) {
-        return next();
+
+  const retryStream = through.obj();
+  let chunkTransformer;
+
+  const makeNewRequest = () => {
+    let lastRowKey = chunkTransformer ? chunkTransformer.lastRowKey : '';
+    chunkTransformer = new ChunkTransformer({decode: options.decode});
+    var grpcOpts = {
+      service: 'Bigtable',
+      method: 'readRows',
+      retryOpts: {
+        currentRetryAttempt: numRequestsMade,
+      },
+    };
+
+    var reqOpts = {
+      tableName: this.id,
+      objectMode: true,
+    };
+    if (lastRowKey) {
+      const lessThan = (lhs, rhs) => {
+        return (
+          Mutation.convertToBytes(lhs).compare(Mutation.convertToBytes(rhs)) ===
+          -1
+        );
+      };
+      const greaterThan = (lhs, rhs) => {
+        return lessThan(rhs, lhs);
+      };
+      const greaterThanOrEqualTo = (lhs, rhs) => {
+        return !lessThan(rhs, lhs);
+      };
+
+      if (ranges.length === 0) {
+        ranges.push({
+          start: {
+            value: lastRowKey,
+            inclusive: false,
+          },
+        });
+      } else {
+        // Readjust and/or remove ranges based on previous valid row reads.
+
+        // Iterate backward since items may need to be removed.
+        for (let index = ranges.length - 1; index >= 0; index--) {
+          const range = ranges[index];
+          const normalizeRangeValue = rangeSide => {
+            return is.object(rangeSide) ? rangeSide.value : rangeSide;
+          };
+          const startValue = normalizeRangeValue(range.start);
```

Comment:
I think we can avoid the inline function `normalizeRangeValue` if we just:

```js
const startValue = range.start.value || range.start;
const endValue = range.end.value || range.end;
```

Reply:
It's possible for `range.start.value` to be an empty string which would cause the value to be an object. Probably not the intended behavior. Is something like this better than the current version?

```js
const startValue = is.object(range.start) ? range.start.value : range.start;
const endValue = is.object(range.end) ? range.end.value : range.end;
```

Reply:
Sure, that LGTM.